model/version_db.go

```diff
@@ -64,27 +64,29 @@ func VersionByIds(ids []string) db.Q {
 // All is a query for all versions.
 var VersionAll = db.Query(bson.D{})
 
-// FindVersionByLastKnownGoodConfig filters on versions with valid (i.e., have no errors) config for the given
-// project. Does not apply a limit, so should generally be used with a findOneRepoRefQ.
+// FindVersionByLastKnownGoodConfig filters on versions with valid (i.e., have no errors) config for the given project.
 func FindVersionByLastKnownGoodConfig(projectId string, revisionOrderNumber int) (*Version, error) {
+	retryLimit := 50
 	q := bson.M{
 		VersionIdentifierKey: projectId,
-		VersionRequesterKey: bson.M{
-			"$in": evergreen.SystemVersionRequesterTypes,
-		},
-		VersionErrorsKey: bson.M{
-			"$exists": false,
-		},
+		VersionRequesterKey:  evergreen.RepotrackerVersionRequester,
 	}
-
 	if revisionOrderNumber >= 0 {
 		q[VersionRevisionOrderNumberKey] = bson.M{"$lt": revisionOrderNumber}
 	}
-	v, err := VersionFindOne(db.Query(q).Sort([]string{"-" + VersionRevisionOrderNumberKey}))
-	if err != nil {
-		return nil, errors.Wrapf(err, "Error finding recent valid version for '%s'", projectId)
+	for i := 0; i < retryLimit; i++ {
+		v, err := VersionFindOne(db.Query(q).Sort([]string{"-" + VersionRevisionOrderNumberKey}))
```

Comment:
I don't think .Sort works on FindOne(). What is this trying to do?

Reply:
This didn't change in this PR. This is the same as writing `db.<coll>.findOne(<query>).sort(<query>)` or  `db.<coll>.find(<query>).sort(<query>).limit(1)`. It's trying to find the document with the greatest `VersionRevisionOrderNumberKey`.

Reply:
Based on what I'm reading online its seems like `db.<coll>.findOne(<query>).sort(<query>)` would not work, but `db.<coll>.find(<query>).sort(<query>).limit(1)` would. However, if we know that this has been working till now, then maybe what I'm reading is outdated.

Reply:
Here sort is setting the q.sort value for the query: https://github.com/evergreen-ci/evergreen/blob/main/db/query.go#L51

So that later this property can be used in FindOneQ: https://github.com/evergreen-ci/evergreen/blob/main/db/query.go#L82

This allows us to retrieve a sorted response with most recent version revision number first and take the first element, hence Limit(1) that is applied.

Reply:
aaah that makes sense, thank you for explaining that!